expand.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"io"
+
+	"golang.org/x/text/transform"
+)
+
+// expandReadCloser implements io.ReadCloser
+type expandReadCloser struct {
+	io.Reader
+	io.Closer
+}
+
+func newExpandReadCloser(rc io.ReadCloser, expandFunc func(string) (string, error)) io.ReadCloser {
+	return &expandReadCloser{
+		transform.NewReader(
+			rc,
+			&expandTransformer{expand: expandFunc}),
+		rc,
+	}
+}
+
+// expandTransformer implements transform.Transformer
+type expandTransformer struct {
+	transform.NopResetter
+	expand func(string) (string, error)
+}
+
+// First char of shell variable may be [a-zA-Z_]
+func isShellNameFirstChar(c byte) bool {
```

Comment:
Looking at the implementation of os.Expand, it looks like it supports `$1` and `$1foo`. Should we replicate that behavior?

https://play.golang.org/p/3g_9wwuQU_

Reply:
Well, `$1foo` would be parsed as `$1` followed by the literal characters `foo`.  In shell context $1, $2, etc refer to parameter arguments,  $* refers to the concatenated parameter arguments, $@ refers to the quoted parameter arguments, $$ refers to the process id, etc.  They all have specific meaning in a shell context which don't really seem to make sense here.  I'd stick strictly to shell variable names which have the form "[a-zA-Z_][a-zA-Z0-9_]*".  But I don't feel too strongly about it.

Of course, the code I've uploaded doesn't strictly parse variables wrapped in curly braces, so it's possible that someone could use ${1} as a variable name.  Not that that's a good thing, just pointing it out.

Reply:
I don't feel strongly about whether we should support `$1` or not. It won't be a valid environment variable for sure. I'm happy to leave that as-is.